jumanji/environments/routing/connector/generator.py

```diff
@@ -106,3 +125,444 @@ def __call__(self, key: chex.PRNGKey) -> State:
+        keys = jax.random.split(key, num=self.num_agents)
+
+        # Randomly select action for each agent
+        actions = jax.vmap(self._select_action, in_axes=(0, None, 0))(
+            keys, grid, agents
+        )
+
+        # Step all agents at the same time (separately) and return all of the grids
+        new_agents, grids = jax.vmap(self._step_agent, in_axes=(0, None, 0))(
+            agents, grid, actions
+        )
+
+        # Get grids with only values related to a single agent.
+        # For example: remove all other agents from agent 1's grid. Do this for all agents.
+        agent_grids = jax.vmap(get_agent_grid)(agent_ids, grids)
+        joined_grid = jnp.max(agent_grids, 0)  # join the grids
+
+        # Create a correction mask for possible collisions (see the docs of `get_correction_mask`)
+        correction_fn = jax.vmap(get_correction_mask, in_axes=(None, None, 0))
+        correction_masks, collided_agents = correction_fn(grid, joined_grid, agent_ids)
+        correction_mask = jnp.sum(correction_masks, 0)
+
+        # Correct state.agents
+        # Get the correct agents, either old agents (if collision) or new agents if no collision
+        agents = jax.vmap(
+            lambda collided, old_agent, new_agent: jax.lax.cond(
+                collided,
+                lambda: old_agent,
+                lambda: new_agent,
+            )
+        )(collided_agents, agents, new_agents)
+        # Create the new grid by fixing old one with correction mask and adding the obstacles
+        return agents, joined_grid + correction_mask
+
+    def _initialise_agents(
+        self, key: chex.PRNGKey, grid: chex.Array
+    ) -> Tuple[chex.Array, Agent]:
+        """Initialises agents using random starting point and places heads on grid.
+
+        Args:
+            key: random key.
+            grid: empty grid.
+
+        Returns:
+            Tuple of grid with populated starting points and agents initialised with
+            the same starting points.
+        """
+        starts_flat = jax.random.choice(
+            key=key,
+            a=jnp.arange(self.rows * self.cols),
+            shape=(self.num_agents,),
+            # Start positions for all agents
+            replace=False,
+        )
+
+        # Create 2D points from the flat arrays.
+        starts = jnp.divmod(starts_flat, self.cols)
+        # Fill target with default value as targets will be assigned aftert random walk
+        targets = jnp.full((2, self.num_agents), -1)
+
+        # Initialise agents
+        agents = jax.vmap(Agent)(
+            id=jnp.arange(self.num_agents),
+            start=jnp.stack(starts, axis=1),
+            target=jnp.stack(targets, axis=1),
+            position=jnp.stack(starts, axis=1),
+        )
+        # Place agent heads on grid
+        grid = jax.vmap(self._place_agent_heads_on_grid, in_axes=(None, 0))(
+            grid, agents
+        )
+        grid = grid.max(axis=0)
```

Comment:
Since the output of the vmap is multiple grids (one for each agent) this is to combine all the information together into one grid to move forward with. Alternatively we could also sum the grids if that's preferred?

Reply:
I renamed the vmap output to grids to disambiguate, does that work?